src/ptvsd/_vendored/pydevd/_pydevd_bundle/pydevd_process_net_command_json.py

```diff
@@ -268,5 +272,126 @@ def on_setbreakpoints_request(self, py_db, request):
+        Note that in the DAP the variables reference requires a unique int... the way this works for
+        pydevd is that an instance is generated for that specific variable reference and we use its
+        id(instance) to identify it to make sure all items are unique (and the actual {id->instance}
+        is added to a dict which is only valid while the thread is suspended and later cleared when
+        the related thread resumes execution).
+
+        see: SuspendedFramesManager
+
+        :param VariablesRequest request:
+        '''
+        arguments = request.arguments  # : :type arguments: VariablesArguments
+        variables_reference = arguments.variablesReference
+
+        thread_id = py_db.suspended_frames_manager.get_thread_id_for_variable_reference(
+            variables_reference)
+        if thread_id is not None:
+            self.api.request_get_variable_json(py_db, request, thread_id)
+        else:
+            variables = []
+            body = VariablesResponseBody(variables)
+            variables_response = pydevd_base_schema.build_response(request, kwargs={'body':body})
+            return NetCommand(CMD_RETURN, 0, variables_response.to_dict(), is_json=True)
+
+    def on_setvariable_request(self, py_db, request):
+        arguments = request.arguments  # : :type arguments: SetVariableArguments
+        variables_reference = arguments.variablesReference
+
+        thread_id = py_db.suspended_frames_manager.get_thread_id_for_variable_reference(
+            variables_reference)
+        if thread_id is not None:
+            self.api.request_change_variable_json(py_db, request, thread_id)
+        else:
+            body = SetVariableResponseBody('')
+            variables_response = pydevd_base_schema.build_response(
+                request,
+                kwargs={
+                    'body':body,
+                    'success': False,
+                    'message': 'Unable to find thread to evaluate variable reference.'
+            })
+            return NetCommand(CMD_RETURN, 0, variables_response.to_dict(), is_json=True)
+
+    def on_modules_request(self, py_db, request):
+        modules_manager = py_db.cmd_factory.modules_manager  # : :type modules_manager: ModulesManager
+        modules_info = modules_manager.get_modules_info()
+        body = ModulesResponseBody(modules_info)
+        variables_response = pydevd_base_schema.build_response(request, kwargs={'body':body})
+        return NetCommand(CMD_RETURN, 0, variables_response.to_dict(), is_json=True)
+
+    def on_source_request(self, py_db, request):
+        '''
+        :param SourceRequest request:
+        '''
+        source_reference = request.arguments.sourceReference
+        content = ''
+        if source_reference != 0:
+            server_filename = pydevd_file_utils.get_server_filename_from_source_reference(source_reference)
+            if os.path.exists(server_filename):
+                with open(server_filename, 'r') as stream:
```

Comment:
This needs to handle any I/O or permission errors, and a potential race against the file being deleted right after the check above.

Reply:
Agreed. This should be properly tested too to return proper errors on those cases, so, created https://github.com/Microsoft/ptvsd/issues/1173 to address this.